### DISCUSSION TITLE
1. MSA란??

### DISCUSSION THREAD

**Discussion by JoisFe** [cs-collections/cs-collections-discussion#14] [j. MSA] [answered]
# MSA란??

## Architecture
- 모든 시스템은 하나 이상의 컴포넌트로 구성.
- 즉 API 컴포넌트나 저장소 컴포넌트처럼 각자 역할이 분리.

### Monolithic System Architecture : 서비스 기능을 하나의 API 컴포넌트에서 처리하는 구조
- 즉 하나의 시스템이 서비스 전체 기능을 처리하도록 설계

### 장점
- 하나의 WAS(Web Application Server)에서 모든 기능을 처리하도록 구성
- 데이터를 저장하기 위해 하나의 데이터 저장소를 사용

![image](https://user-images.githubusercontent.com/90208100/208831471-0cbce9e9-55b6-4ca7-909b-9dcdee200cf7.png)
<그림 참고 : https://velog.io/@hayoungishere/MSA%EB%A5%BC-%EC%95%8C%EC%95%84%EB%B3%B4%EC%9E%90>

위 특징으로 인해
1. 구조가 매우 간단
2. 구조가 간단하기에 시스템 운영과 개발이 편리한 장점
3. 애플리케이션이 1개 이므로 code base도 하나면 충분
4. MSA와 다르게 네트워크로 인한 지연이나 데이터 유실 걱정할 필요 없음
5. 시스템 장애 혹은 기능에 버그가 있다면 개발자는 하나의 애플리케이션에서 원인을 파악하면 됨
6. 데이터 저장소가 1개 이므로 RDB의 트랜잭션 기능을 쉽게 사용
7. 테스트 환경을 쉽게 구성
8. 장애에 견고

### 단점
- 서비스에 대한  사용자의 요청이 늘어난 상황이 생겼다면, 감당하기 위해 시스템 확장을 해야하고 시스템 안정성을 위해 고가용성(HA: High Availability)를 확보를 해야한다.

1. 하나의 애플리케이션 서버에서 여러 기능을 제공하므로 서비스 기능이 많아지면 더욱 복잡. (스파게티 코드)
2. 서비스 규모가 커지고 제공하는 기능도 많아지면서 개발자가 늘고 GIT 브랜치가 많아지는 등 관리하는데 서로 충돌하는 코드를 수정하는 경우가 많아짐
3.  하나의 code base에 여러 기능이 뒤 섞여 있어 클라이언트 코드 버그를 수정할 경우 서버를 재시작 해야하는 등의 불편함이 존재.
4. 위의 상황을 해결하기 위해 load balancer를 이용하여 scale-out한 경우가 존재하는데 load balancer를 이용한 방법으로 확장해도 성능에 한계가 존재함.
5. 모놀리식 아키텍처에서는 스케일 아웃하는 것이 비효율적일 수 있음 (특정 기능에 대한 요청만 많지만 하나의 WAS에 모든 기능이 처리되어 있으므로 WAS 1개를 통으로 scale out 해야 함)

### MSA : 분산 처리 시스템 중 하나
- 기능 위주로 나뉜 여러 애플리케이션 존재
- 각각 독립된 데이터 저장소를 사용
- 웹이나 App 클라이언트에 일관된 형태의 API로 제공

![image](https://user-images.githubusercontent.com/90208100/208833837-c67d8c15-ffaa-4f3e-9733-0740d6f6605b.png)
<그림 참고 : https://velog.io/@hayoungishere/MSA%EB%A5%BC-%EC%95%8C%EC%95%84%EB%B3%B4%EC%9E%90>

### Microservice : 기능별로 쪼개진 작은 서비스 혹은 시스템

- MSA의 microservice 들은 각각 다른 microservice에서 독립적으로 구성되어야 함.
- microservice들은 느슨하게 결합 (다른 서비스에 의존성 최소화) 되어 각각 독립성을 갖도록 해야함
- microservice 마다 각각 독립된 데이터 저장소가 필요 
- microservice들은 기능과 성격에 맞게 잘 분리 되어야 함.

### 장점
1. 독립성
- 하나의 마이크로서비스는 하나의 비즈니스 기능을 담당 -> 다른 마이크로서비스와 간섭이 최소화
- 하나의 마이크로서비스는 하나의 데이터 저장소를 가짐 -> 데이터 간섭에도 자유로움
2. 대용량 데이터를 저장하고 처리하는데 비교적 자유로움
- RDB의 스케일 아웃 쉽지 않음 --> 데이터 샤딩
- 마이크로서비스는 독립된 데이터 저장소를 가지고 있기에 대용량 데이터를 마이크로서비스마다 나누어 저장 가능
- 샤딩이나 서비스에 적합한 캐시를 마이크로서비스 단위로 각각 적용 가능
3. 시스템 장애에 견고
- 마이크로서비스는 서로 느슨하게 결합되어 있기에 각각 독립되어 있고 이로 인해 서로 간에 미치는 영향이 적다
- 모놀리식은 애플리케이션 서버가 고가용성 방식과 스케일 아웃을 이용해 시스템 장애를 대비하지만 애플리케이션 서버 중 하나에만 문제 발생시 해당 장비는 격리되고 나머지 서버들이 서비스를 제공. 하지만 다수의 장비에 문제가 발생 혹은 버그가 발생 시 시스템 전체로 영향이 확산됨
- 마이크로서비스들이 느슨하게 결합되어 있기에 하나의 마이크로서비스에 장애나 버그가 발생하더라도 다른 마이크로서비스는 이상 없이 서비스 가능
4. 마이크로서비스와 클라우드 서비스에서 제공하는 몇몇 기능을 결합하여 애플리케이션은 탄력 회복성 (resilience)를 가짐
- 탄력 회복성 : 애플리케이션 서버에 장애가 발생하면 새로운 컴퓨팅 자원을 추가하여 빠른 시간 내에 서비스를 다시 제공하는 것
5. 서비스 배포 주기가 빠름
- 모놀리식 경우 모든 기능이 하나의 code base에서 개발되기에 일반적으로 배포 일정을 정하고 그 기간 동안 개발된 모든 기능을 한번에 배포 (서비스 마다 모두 개발 시간이 같지 않으므로 가장 긴 시간이 필요한 기능에 배포 일정을 맞춤) --> 비효율적
- 마이크로서비스 환경에서는 모든 기능이 분리되어 있으므로 필요한 기능만 먼저 배포 가능
6. 마이크로서비스 단위로 확장할 수 있어 서비스 전체적으로 확장성이 좋아짐, 거기에 클라우드 시스템과 결합하면 시스템 컴포넌트들을 동적으로 확장할 수 있음
- 모놀리식 경우 서비스를 확장할 경우 모든 기능을 확장해야하는 비효율이 발생
7. 사용자 반응에 민첩하게 대응 가능.
- 새로운 서비스를 마이크로서비스로 분리하여 설계하고 시스템에 포함하면 사용자 반응에 따라 시스템을 고도화하거나 빠르게 시스템에서 제외할 수 있음.

### 단점
1. 개발하기 어려운 아키텍처
- 프로세스가 다른 마이크로서비스를 개발할 때 고려해야할 상황이 많음
- 특히 가장 고려해야 하는 부분이 분리된 데이터, 네트워크를 이용한 데이터 통합 임
- 데이터 저장소가 분리되어 있기에 RDB의 매우 큰 장점 중 하나인 트랜잭션을 사용할 수 없음 (데이터 정합성이 맞지 않거나 중복된 데이터 발생 가능성이 큼)
- 분산 트랜잭션을 사용할 수 있겠지만 분산 트랜잭션은 시스템 전체의 리소스를 많이 사용하므로 권장하지는 않는다고 함.
- 데이터를 통합하기 위해서는 표준화된 인터페이스로 데이터를 처리해야함
- 네트워크는 신뢰할 수 없고 커넥션을 맺는 비용이 비쌈. (네트워크는 언제나 장애가 발생 가능, 패킷은 언제든 누락 가능) -> 요청한 응답을 받지 못하는 상황이 올 수 있고 네트워크 지연이 발생할 수 있음
2. 운영하기 어려운 아키텍처
- 요청에 대한 처리를 하기 위해여러 마이크로서비스의 통합된 데이터가 필요한 경우가 많은데 이 경우 너무 복잡한 상황이 발생
- 요청이 정상적으로 처리되지 않았을 시 어느 부분에 에러가 발생했는지 찾기도 쉽지 않음
- 데이터가 분산되어 있기에 분산 트랜잭션을 사용하지 않는다면 데이터 일관성을 유지하기 어려움.
3. 설계 하기 어려운 아키텍처
- 잘 설계된 MSA 시스템은 복잡한 비즈니스 로직을 개발해서 보다 쉽게 유지 보수 가능
- 잘못된 방향으로 설계된 경우 유지 보수 하는데 더 많은 시간과 비용이 소모
- 마이크로서비스의 관계가 명확하게 구분되지 못한다면 각 마이크로서비스가 관리하는 데이터들이 중복될 수 있음. (데이터의 파편화 현상 때문에 서비스 전체적으로 일관성 있는 데이터를 유지하기 어려움)
- 마이크로서비스간의 관계가 명확하지 못하면 이들 사이에 과도한 네트워크 통신이 발생 -> 전체적인 성능 저하.
- 어떻게 어느 정도 크기로 서비스를 나누어야 하는지 기준을 만드는 과정이 매우 까다로움
4. MSA로 설계된 서비스 운영에는 여러 가지 자동화된 시스템이 필요
- MSA를 도입하는 가장 큰 이유가 빠른 서비스 개발과 운영, 대규모 서비스를 처리하기 위해서
- 작게 분리된 마이크로서비스는 적은 인원으로 빠르게 개발 가능
- 또한 많은 인스턴스에 빠르게 배포하기 위해 CI/CD 시스템이 필요
- 수많은 인스턴스를 사람의 손으로 일일이 모니터링 하기 불가능하므로 모니터링 시스템 또한 필요.
- 로그를 통합해서 검색해주는 시스템 또한 필요
5. 개발자의 기술력이 좋아야 함
6. 연쇄 장애 (Chain of Failure) 가 발생할 수 있음
- 통신하는 다른 시스템에서 오류가 발생할때 플랫폼으로 전이될 수 있음
- 플랫폼에서 보낸 요청에 다른 시스템이 제시간에 응답하지 않을 시 플랫폼이 동시에 많은 양의 요청을 받았을때 OS thread와 같은 주요 자원이 부족
- 이에 따라 플랫폼의 컴포넌트가 지연되거나 중단.
- 플랫폼 내부 통신은 대부분 동기식이므로 한 컴포넌트의 중단은 연속적인 오류를 유발
- 즉 중단된 컴포넌트의 클라이언트도 곧 중단 (연쇄 장애)

## 참고 자료
1. 스프링 부트로 개발하는 MSA 컴포넌트 - 김병부 지음, 길벗 출판
2. 스프링으로 하는 마이크로 서비스 구축 - 매그너스 라슨 지음, 에이콘 출판

**Comment by ahngilwoong** [comment, marked as answer]
잘 봤습니다 뭐 하나 아무거나 질문하려고 불 키고 봤는데 잘 안떠오르네요.
4. MSA로 설계된 서비스 운영에는 여러 가지 자동화된 시스템이 필요 에
대구모라고 하셨는데 대규모로 수정 부탁드립니다.

**Reply by JoisFe** [reply]
감사합니다. 수정하겠습니다.
